pkg/cache/v3/linear.go

```diff
@@ -488,104 +595,104 @@ func (cache *LinearCache) CreateWatch(request *Request, sub Subscription, value
 	}, nil
 }
 
-func (cache *LinearCache) nextSotwWatchID() int64 {
-	next := atomic.AddInt64(&cache.sotwWatchCount, 1)
-	if next < 0 {
-		panic("watch id count overflow")
-	}
-	return next
-}
-
 // Must be called under lock
-func (cache *LinearCache) removeWatch(watchID int64, sub Subscription) {
+func (cache *LinearCache) removeWatch(watchID uint64, sub Subscription) {
 	// Make sure we clean the watch for ALL resources it might be associated with,
 	// as the channel will no longer be listened to
 	for resource := range sub.SubscribedResources() {
-		resourceWatches := cache.watches[resource]
-		delete(resourceWatches, watchID)
-		if len(resourceWatches) == 0 {
-			delete(cache.watches, resource)
+		resourceWatches := cache.resourceWatches[resource]
+		delete(resourceWatches.sotw, watchID)
+		if resourceWatches.empty() {
+			delete(cache.resourceWatches, resource)
 		}
 	}
 }
 
 func (cache *LinearCache) CreateDeltaWatch(request *DeltaRequest, sub Subscription, value chan DeltaResponse) (func(), error) {
+	if request.GetTypeUrl() != cache.typeURL {
```

Comment:
Most changes in delta is to align with `CreateWatch`. The content is now the same except for the potential versions initialization